rclcpp/include/rclcpp/node.hpp

```diff
@@ -1072,6 +1072,58 @@ class Node : public std::enable_shared_from_this<Node>
   bool
   assert_liveliness() const;
 
+  /// Return the topic endpoint information about publishers on a given topic.
+  /**
+   * The returned parameter is a list of topic endpoint information, where each item will contain
+   * the node name, node namespace, topic type, endpoint type, topic endpoint's GID, and its QoS
+   * profile.
+   *
+   * When the `no_mangle` parameter is `true`, the provided `topic_name` should be a valid topic
+   * name for the middleware (useful when combining ROS with native middleware (e.g. DDS) apps).
+   * When the `no_mangle` parameter is `false`, the provided `topic_name` should follow
+   * ROS topic name conventions.
+   *
+   * `topic_name` may be a relative, private, or fully qualified topic name.
+   * A relative or private topic will be expanded using this node's namespace and name.
+   * The queried `topic_name` is not remapped.
+   *
+   * \param[in] topic_name the topic_name on which to find the publishers.
+   * \param[in] no_mangle if `true`, `topic_name` needs to be a valid middleware topic name,
+   *   otherwise it should be a valid ROS topic name. Defaults to `false`.
+   * \return a list of TopicEndpointInfo representing all the publishers on this topic.
+   * \throws InvalidTopicNameError if the given topic_name is invalid.
+   * \throws std::runtime_error if internal error happens.
+   */
+  RCLCPP_PUBLIC
+  std::vector<rclcpp::TopicEndpointInfo>
+  get_publishers_info_by_topic(const std::string & topic_name, bool no_mangle = false) const;
```

Comment:
These two methods were not also added to the `LifecycleNode` class.

This keeps coming up and so to avoid it we could consider leaving these methods off of `Node` and instead have free functions that take the node as a template argument instead, something like:

```c++
auto publisher_infos = get_publishers_info_by_topic(node, "topic_name");
```

Where `node` could be a `rclcpp::Node` or a `rclcpp_lifecycle::LifecycleNode`.

Reply:
Either way, this pull request should make sure `Node` and `LifecycleNode` can both be used the same way (either each having these methods, or the free function).

Reply:
I think this is actually simpler than that. There's no implementation in `Node::get_publishers_info_by_topic()`, it is just a wrapper around `node_graph_->get_publishers_info_by_topic()`. The same wrapper can be added to `LifecycleNode`.

Reply:
> I think this is actually simpler than that. There's no implementation in `Node::get_publishers_info_by_topic()`, it is just a wrapper around `node_graph_->get_publishers_info_by_topic()`. The same wrapper can be added to `LifecycleNode`.

No doubt it can be added, my point was simply that if we standardized on a templated free function that takes a node instead, we'd not need to have the "wrapper" methods for each class. Keeping them in sync has not been trivial.

But until we do an API review and make a decision on that point, adding it to the `LifecyleNode` manually is fine.